Fix for mute not working

bmaupin opened the pull request:
Fixes https://github.com/danigm/spotify-ad-blocker/issues/33

It seems Spotify is now resetting the stream volume while ads are playing. Adds handlers to all spotify streams so if the volume is changed while an ad is playing it can be re-muted as necessary.

The logic to determine whether a mute or unmute should happen has been pulled out of the mute() and unmute() functions since it now depends on whether they're being called because the track has changed (`update()`) or the volume has changed.

bmaupin committed:
feat: Listen for stream volume changes

... and re-mute as needed.

Fixes https://github.com/danigm/spotify-ad-blocker/issues/33

bmaupin committed:
fix: Check if new stream needs to be muted

... right away as a new stream might be created while an ad is playing, e.g. if a video ad is played, the audio stream will be removed. Then if an audio ad is played immediately afterward, a new audio stream will be created.

Also set stream handler variables back to 0 when they're disconnected.

bmaupin commented:
@danigm I've been testing this and I thought it was ready, but something weird happened today. At some point Spotify was playing a track (not an ad) with the muted volume. This can occur while I'm testing for a number of reasons (I'll restart X11 when testing to pick up changes to the extension, or X11 can crash sometimes) and normally it's not an issue: I just manually increase the volume.

However, today I noticed that even when I increased the volume to full, it kept getting changed to its previous level. The extension won't do this (if an ad isn't playing) so I'm pretty sure Spotify was resetting the volume. My best guess is that Spotify must be getting the stream volume at some point (maybe when it's first opened) and that's the volume it uses when it resets the volume.

This seems to be reflected in [this comment](https://github.com/danigm/spotify-ad-blocker/issues/33#issuecomment-3191543111):

> I realized, that setting volume in spotify also changed behaviour. When I set a lower volume in Gnome's audio mixer, it is reset to 100% when the next track starts. This is regardless of it being an ad or music.

To fix this, I had to set the spotify stream volume to 100%, then completely close and reopen Spotify. As best as I can tell it's detecting the stream volume at startup, so as long as it's started up at 100%, that's the volume it will use.

So I'm trying to figure out how to account for this. I have these ideas right now:

A. The simplest approach I can think of would be to ignore user-defined volume. Up until this point, in the code I've been trying to account for the possibility that the user may want to change the volume of the spotify stream specifically. However, if Spotify is overriding the stream volume when the track is changed (whether it's an ad or not), it means that Spotify itself is overriding any user-defined volume with whatever stream volume was set when it was first launched.

If we take this approach, we never need to store previous volume. We can always just set it to 100% when unmuting or 100% times `ad-volume-percentage` when muting. Then it should be pretty easy to add logic to detect deviations from either of these two volume levels.

B. Alternatively we could try to differentiate between user set volume and Spotify set volume. For example, we could detect any volume change within a certain time window of a track change as a volume change initiated by Spotify. But I feel like this would be complicated and error-prone and may lead to unexpected behaviour, e.g. if the user changes the volume close to a track change. On the other hand, ignoring user-defined volume altogether (option A) might also be unexpected behaviour.

What do you think? Any other ideas?

danigm reviewed (approved):
LGTM, Great work, very clever to track volume changes and mute again if needed. It's clear that spotify is changing the changing the volume so this change is great to get any change and make sure that it's muted when it should be.

bmaupin commented:
@danigm Okay, great, I'll start working on option A and testing it. I like the simplicity of it too.

I'm not sure it's worth adding now, but I did think of a possible solution to the problem of user-defined stream volume. If users don't like not being able to set the stream volume themselves, we could add a new configuration setting. Since we already have _Volume percentage for ads_, we could also add _Volume percentage for music_, which would default to 100. That would allow users to override the stream volume.

But I thought we could wait to add that if it becomes an issue and go with the simplest approach for now (option A). I'm planning on adding it to this pull request. Is that okay or would you like a separate pull request for that?

bmaupin committed:
feat: Handle only two volume levels:

... muted or unmuted. Since Spotify is modifying the stream volume, it's no longer feasible to distinguish between volume set by the user versus volume set by Spotify.

bmaupin committed:
feat: Apply unmute to volume changes as well

If the stream volume when Spotify is first opened is lower than max volume, Spotify will lower the stream volume even when music is playing, so we may need to unmute as well

bmaupin committed:
feat: Don't delay unmute on volume change

Unmute was previously delayed because it was only triggered by a track change, and sometimes that signal would be emitted before the new track would start playing. This does not seem to be the case with the volume change signal so we should unmute right away.

danigm commented:
@bmaupin I'm okay with the current state. We can just mute/unmute and do not track the specific spotify stream volume for now, just not supported until we find a way to do it correctly without spotify app messing around.

bmaupin commented:
Sounds good. I'll test for a couple more days or so and then mark this as ready if I don't run into any more issues.

bmaupin commented:
I'm going to go ahead and mark this as ready. I still feel a tiny bit iffy about the UX only because it's slightly inconsistent. But if it's ever an issue I think it wouldn't be too much work to make it consistent, it would just be a matter of deciding on the desired behaviour.

For example, one approach would be when muting, only check if the volume is the unmuted volume (e.g. max) and when unmuting only check if the volume is the muted volume. This would allow the user to adjust the volume when a track or ad is playing and the volume would only be reset on the next track/ad or if it happens to hit one of those levels. This would be pretty straightforward to implement, although slightly more complex than the current logic, which more or less is only tracking max volume:

- move `muted()` logic into `shouldMute`
- `shouldMute` will only mute if the volume is max volume
- `shouldUnmute` will only mute if the volume is muted volume
  - we could check this either by having a separate function that gets mutedVolume (probably cleanest?) or checking muted volume +/- 1 (to account for the conversion to integer rounding up or down)
